File: mobile/welcome-screen/ios-app/ShowUpWelcome/NotificationAccess.swift
//
//  NotificationAccess.swift
//  ShowUpWelcome · the notification permission status, and the case that skips the ask (SHOWUP-162)
//
//  ───────────────────────────────────────────────────────────────────────────
//  FOUR VALUES, ONE VOCABULARY, BOTH PLATFORMS
//  ───────────────────────────────────────────────────────────────────────────
//
//  `enums.json` §24 defines the set and the ticket uses it verbatim: `not_determined · granted ·
//  denied · restricted`. Not spelled per-platform — iOS's `authorizationStatus` and Android's
//  rationale flag both map onto these four — and `limited` is deliberately absent, because it is a
//  photo-library concept that lives on `permission_result` alone.
//
//  `denied` COVERS PERMANENT DENIAL, and on iOS that is the only kind: the system dialog is shown
//  once per install, so a denial here can never be re-asked. That is exactly why a denial is
//  recovered on Stay reachable (10), which deep-links to Settings, and not on this screen.
//
//  ───────────────────────────────────────────────────────────────────────────
//  WHAT iOS RETURNS THAT ANDROID DOES NOT
//  ───────────────────────────────────────────────────────────────────────────
//
//  `.provisional` and `.ephemeral` both exist and neither is a status this product wants.
//
//  PROVISIONAL IS MAPPED TO `granted` AND NEVER REQUESTED. The ticket forbids asking for it — it
//  delivers quietly to Notification Centre with no banner and no sound, "which is exactly what a
//  date reminder must not be". But if the app were ever launched with it already in place, the
//  status IS determined and the ask would be a dead screen, so it skips like any other answer.
//
//  `.ephemeral` is an App Clip state this app cannot be in; it maps the same way for safety
//  rather than being left to a `default` that could silently become `not_determined`.
//

import Foundation
import UserNotifications

/// The OS's answer, in `enums.json` §24's vocabulary.
enum NotificationPermission: String, Sendable {
    case notDetermined = "not_determined"
    case granted
    case denied
    case restricted

    /// What goes on the wire. The names are never typed at a call site.
    var trackingValue: String { rawValue }

    /// Whether the OS has an answer. The ask exists only for the one status that does not.
    var isDetermined: Bool { self != .notDetermined }
}

/// Should the ask be shown at all?
///
/// The single decision behind the skip case, as a free function so the host can take it before the
/// screen is pushed and a test can take it with no device.
func shouldShowAsk(_ status: NotificationPermission) -> Bool { !status.isDetermined }

/// Maps `UNAuthorizationStatus` onto §24. A pure function, so every branch is testable.
func notificationPermission(for status: UNAuthorizationStatus) -> NotificationPermission {
    switch status {
    case .notDetermined: return .notDetermined
    case .denied: return .denied
    // See the header: neither is requested, and both are answers.
    case .authorized, .provisional, .ephemeral: return .granted
    @unknown default:
        // A status this build has never heard of is an ANSWER, not an absence. Treating it as
        // `not_determined` would show an ask whose sheet may never appear, and this screen has no
        // way out but that sheet.
        return .granted
    }
}

/// Reads the status. Re-read on every foreground — never remembered.
///
/// §24: a permission "is the device's answer and is re-read on every foreground rather than
/// remembered". A cached value is wrong the moment the user changes it in Settings.
protocol NotificationAccessReading: Sendable {
    func read() async -> NotificationPermission
}

/// The real reader.
struct UNNotificationAccess: NotificationAccessReading {
    func read() async -> NotificationPermission {
        let settings = await UNUserNotificationCenter.current().notificationSettings()
        return notificationPermission(for: settings.authorizationStatus)
    }
}

/// Answers from memory. Previews, and every test that is about the rules rather than the device.
struct FixedNotificationAccess: NotificationAccessReading {
    let status: NotificationPermission

    init(_ status: NotificationPermission = .notDetermined) { self.status = status }

    func read() async -> NotificationPermission { status }
}

/// Raises the OS sheet.
///
/// SEPARATE FROM THE READER, because the two are different acts with different rules: the status
/// is read on every foreground and the sheet is raised exactly once, on a tap.
protocol NotificationAsking: Sendable {
    /// Requests authorisation. Returns whether it was granted.
    func request() async -> Bool
}

struct UNNotificationAsk: NotificationAsking {
    func request() async -> Bool {
        // Alert, sound and badge — the three a date reminder needs. NOT `.provisional`: see the
        // file header and the ticket, which rules it out in as many words.
        let granted = try? await UNUserNotificationCenter.current()
            .requestAuthorization(options: [.alert, .sound, .badge])
        return granted ?? false
    }
}

/// Answers from memory, for tests.
struct FixedNotificationAsk: NotificationAsking {
    let granted: Bool
    func request() async -> Bool { granted }
}

File: mobile/welcome-screen/android-preview-project/app/src/main/java/com/showup/profile/NotificationAccess.kt
/*
 * NotificationAccess.kt
 * ShowUp · the notification permission status, and the two cases that skip the ask (SHOWUP-162)
 *
 * ─────────────────────────────────────────────────────────────────────────────
 * FOUR VALUES, ONE VOCABULARY, BOTH PLATFORMS
 * ─────────────────────────────────────────────────────────────────────────────
 *
 * `enums.json` §24 defines the set and the ticket uses it verbatim: `not_determined · granted ·
 * denied · restricted`. Not spelled per-platform -- iOS's `authorizationStatus` and Android's
 * rationale flag both map onto these four -- and `limited` is deliberately absent, because it is a
 * photo-library concept that lives on `permission_result` alone.
 *
 * `denied` COVERS PERMANENT DENIAL. The distinction between "denied once" and "denied for good" is
 * the platform's rationale flag, and §24 is explicit that it is a BEHAVIOUR branch rather than a
 * fifth value. Nothing on this screen branches on it: a denial is recovered on Stay reachable (10)
 * and nowhere else.
 *
 * ─────────────────────────────────────────────────────────────────────────────
 * ANDROID 12 AND BELOW REPORTS `granted`, AND THAT IS NOT A SHORTCUT
 * ─────────────────────────────────────────────────────────────────────────────
 *
 * There is no `POST_NOTIFICATIONS` before API 33: notifications are on by default and there is
 * nothing to ask for. §24 says these users are "reported as granted", so that is what this
 * returns -- not `not_determined`, which would be a lie that raises a sheet that cannot exist.
 *
 * The consequence the ticket underlines: those users still need push registration and all five
 * categories. Being skipped by the ASK is not being skipped by the FEATURE.
 *
 * ─────────────────────────────────────────────────────────────────────────────
 * THE SCREEN IS SKIPPED BEFORE IT IS PUSHED, NEVER AFTER IT MOUNTS
 * ─────────────────────────────────────────────────────────────────────────────
 *
 * [shouldShowAsk] is the whole decision and it is a pure function of the status, so the host can
 * answer it before the screen exists. Two cases skip:
 *
 *   Android <= 12        nothing to ask for
 *   already determined   a restore-from-backup, or the app killed while the sheet was up
 *
 * The second is a GUARD, not a designed state. The system dialog is shown once per install, so a
 * screen whose only button raises a dialog that will not appear is a dead end -- and this screen
 * has no skip, no back and no close. Silently means silently: no toast, no confirmation, and no
 * flash of the screen before it navigates.
 */
package com.showup.profile

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.content.ContextCompat

/**
 * The OS's answer, in `enums.json` §24's vocabulary.
 *
 * [trackingValue] is what goes on the wire; the names are never typed at a call site.
 */
enum class NotificationPermission(val trackingValue: String) {
    NotDetermined("not_determined"),
    Granted("granted"),
    Denied("denied"),

    /**
     * Parental controls or MDM.
     *
     * Android has no equivalent and never returns it; iOS does. §24 says it is "treated as denied
     * everywhere in the product", which here means it is a determined status and so it skips the
     * ask -- there is nothing a sheet could change.
     */
    Restricted("restricted"),
    ;

    /** Whether the OS has an answer. The ask exists only for the one status that does not. */
    val isDetermined: Boolean get() = this != NotDetermined
}

/**
 * Reads the status. Re-read on every foreground -- never remembered.
 *
 * §24 again: a permission "is the device's answer and is re-read on every foreground rather than
 * remembered". A cached value is wrong the moment the user changes it in Settings, and this flow
 * has a screen whose entire reason to exist depends on the answer.
 */
interface NotificationAccessReader {
    fun read(): NotificationPermission
}

/**
 * Should the ask be shown at all?
 *
 * The single decision behind both skip cases, as a pure function so the host can take it before
 * the screen is pushed and a test can take it with no device.
 */
fun shouldShowAsk(status: NotificationPermission): Boolean = !status.isDetermined

/**
 * The status, from the three facts Android can report. A pure function, so the branch that
 * actually fires in production is testable with no Context and no device.
 *
 * `hasAsked` is the only way to tell a permission never requested from one refused: Android
 * reports both as not-granted, which is why [PermissionAskLog] exists and is shared with the photo
 * and media steps rather than re-invented here.
 */
fun notificationPermissionFor(
    sdkInt: Int,
    granted: Boolean,
    hasAsked: Boolean,
): NotificationPermission = when {
    // Below 33 there is no runtime permission at all. Granted, per §24 -- and the branch the
    // ticket calls "the one that actually fires in production".
    sdkInt < Build.VERSION_CODES.TIRAMISU -> NotificationPermission.Granted
    granted -> NotificationPermission.Granted
    hasAsked -> NotificationPermission.Denied
    else -> NotificationPermission.NotDetermined
}

/** The real reader. */
class AndroidNotificationAccess(
    private val context: Context,
    /** Injected so a test can exercise the API-level branch that only fires in production. */
    private val sdkInt: Int = Build.VERSION.SDK_INT,
) : NotificationAccessReader {

    override fun read(): NotificationPermission {
        // The context is not touched below 33, where the answer does not depend on it.
        if (sdkInt < Build.VERSION_CODES.TIRAMISU) {
            return notificationPermissionFor(sdkInt, granted = false, hasAsked = false)
        }
        return notificationPermissionFor(
            sdkInt = sdkInt,
            granted = ContextCompat.checkSelfPermission(
                context, Manifest.permission.POST_NOTIFICATIONS,
            ) == PackageManager.PERMISSION_GRANTED,
            hasAsked = PermissionAskLog.hasAsked(
                context, Manifest.permission.POST_NOTIFICATIONS,
            ),
        )
    }
}

/** Answers from memory. Previews, and every test that is about the rules rather than the device. */
class FixedNotificationAccess(
    private val status: NotificationPermission = NotificationPermission.NotDetermined,
) : NotificationAccessReader {
    override fun read(): NotificationPermission = status
}
